src/variable.js

```diff
@@ -188,9 +199,28 @@ function variable_import(remote, name, module) {
 }
 
 function variable_delete() {
+  // Delete any shadow inputs.
+  for (const [name, shadow] of this._shadows.entries()) {
+    shadow.delete();
+    this._shadows.delete(name);
+  }
   return variable_defineImpl.call(this, null, [], noop);
 }
 
+function variable_shadow(name, inputs, definition) {
+  if (arguments.length < 3) definition = inputs, inputs = [];
+  if (!this._shadows.has(name)) this._shadows.set(name, this._module.define(inputs, definition));
```

Comment:
Perhaps shadow variables should be `TYPE_IMPLICIT` rather than `TYPE_NORMAL`, similar to built-in variables from the standard library? In practice it shouldn’t make any difference since shadow variables do not have a name at module scope, but it might still be helpful for debugging. We could also introduce a new `TYPE_SHADOW` if we want to further delineate them… but I’m not sure if that would have other implications.